api/types/saml.go

```diff
@@ -383,165 +374,109 @@ func (o *SAMLConnectorV2) GetTraitMappings() TraitMappingSet {
 		}
 		o.Spec.EntityDescriptor = string(body)
 		log.Debugf("[SAML] Successfully fetched entity descriptor from %q", o.Spec.EntityDescriptorURL)
 	}
 	if o.Spec.EntityDescriptor != "" {
-		metadata := &types.EntityDescriptor{}
+		// metadata is a slimmed down anonymous struct of
+		// github.com/russellhaering/gosaml2/types.EntityDescriptor
+		// to avoid unnecessary imports.
+		metadata := &struct {
+			EntityID         string `xml:"entityID,attr"`
+			IDPSSODescriptor *struct {
+				SingleSignOnServices []struct {
+					XMLName  xml.Name `xml:"urn:oasis:names:tc:SAML:2.0:metadata SingleSignOnService"`
+					Location string   `xml:"Location,attr"`
+				} `xml:"SingleSignOnService"`
+			} `xml:"IDPSSODescriptor,omitempty"`
+		}{}
+
 		err := xml.Unmarshal([]byte(o.Spec.EntityDescriptor), metadata)
 		if err != nil {
-			return nil, trace.Wrap(err, "failed to parse entity_descriptor")
+			return trace.Wrap(err, "failed to parse entity_descriptor")
 		}
 
-		for _, kd := range metadata.IDPSSODescriptor.KeyDescriptors {
-			for _, samlCert := range kd.KeyInfo.X509Data.X509Certificates {
-				certData, err := base64.StdEncoding.DecodeString(strings.TrimSpace(samlCert.Data))
-				if err != nil {
-					return nil, trace.Wrap(err)
-				}
-				cert, err := x509.ParseCertificate(certData)
-				if err != nil {
-					return nil, trace.Wrap(err, "failed to parse certificate in metadata")
-				}
-				certStore.Roots = append(certStore.Roots, cert)
-			}
-		}
 		o.Spec.Issuer = metadata.EntityID
 		if len(metadata.IDPSSODescriptor.SingleSignOnServices) > 0 {
 			o.Spec.SSO = metadata.IDPSSODescriptor.SingleSignOnServices[0].Location
 		}
 	}
 	if o.Spec.Issuer == "" {
-		return nil, trace.BadParameter("no issuer or entityID set, either set issuer as a parameter or via entity_descriptor spec")
+		return trace.BadParameter("no issuer or entityID set, either set issuer as a parameter or via entity_descriptor spec")
 	}
 	if o.Spec.SSO == "" {
-		return nil, trace.BadParameter("no SSO set either explicitly or via entity_descriptor spec")
-	}
-	if o.Spec.Cert != "" {
-		cert, err := tlsca.ParseCertificatePEM([]byte(o.Spec.Cert))
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		certStore.Roots = append(certStore.Roots, cert)
-	}
-	if len(certStore.Roots) == 0 {
-		return nil, trace.BadParameter(
-			"no identity provider certificate provided, either set certificate as a parameter or via entity_descriptor")
+		return trace.BadParameter("no SSO set either explicitly or via entity_descriptor spec")
 	}
+
 	if o.Spec.SigningKeyPair == nil {
 		keyPEM, certPEM, err := utils.GenerateSelfSignedSigningCert(pkix.Name{
 			Organization: []string{"Teleport OSS"},
 			CommonName:   "teleport.localhost.localdomain",
 		}, nil, 10*365*24*time.Hour)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return trace.Wrap(err)
 		}
 		o.Spec.SigningKeyPair = &SigningKeyPair{
 			PrivateKey: string(keyPEM),
 			Cert:       string(certPEM),
 		}
 	}
```

Comment:
This is optional, so just move it to `services.GetSAMLServiceProvider` which is where it was before right?

Reply:
Yes, but there was this block in `CheckAndSetDefaults`
```go
_, err = o.GetServiceProvider(clockwork.NewRealClock())
if err != nil {
	return trace.Wrap(err)
}
```
Maybe I should to add a new `ValidateSamlConnector` function that calls this and `CheckAndSetDefaults` so that issues don't arise during runtime.